galsim/real.py

```diff
@@ -839,3 +908,420 @@ def _parse_files_dirs(file_name, image_dir, dir, noise_dir, sample):
+            self.rng = galsim.BaseDeviate()
+        elif not isinstance(rng, galsim.BaseDeviate):
+            raise TypeError("The rng provided to ChromaticRealGalaxy constructor "
+                            "is not a BaseDeviate")
+        else:
+            self.rng = rng
+        self._rng = self.rng.duplicate()  # This is only needed if we want to make sure eval(repr)
+                                          # results in the same object.
+
+        if _imgs is not None:  # pragma: no cover
+            # Special (undocumented) way to build a ChromaticRealGalaxy without needing the crg
+            # catalog directly by providing the things we need from it.
+            imgs, bands, xis, PSFs = _imgs, _bands, _xis, _PSFs
+            if not hasattr(PSFs, '__iter__'):
+                PSFs = [PSFs]*len(imgs)
+
+            use_index = 0  # For the logger statements below.
+            if logger:
+                logger.debug('ChromaticRealGalaxy %d: Start ChromaticRealGalaxy constructor.',
+                             use_index)
+            self.catalog_files = None
+        else:
+            if real_galaxy_catalogs is None:
+                raise ValueError("No RealGalaxyCatalog(s) specified!")
+
+            # Get the index to use in the catalog
+            if index is not None:
+                if id is not None or random:
+                    raise AttributeError('Too many methods for selecting a galaxy!')
+                use_index = index
+            elif id is not None:
+                if random:
+                    raise AttributeError('Too many methods for selecting a galaxy!')
+                use_index = real_galaxy_catalogs[0].getIndexForID(id)
+            elif random:
+                uniform_deviate = galsim.UniformDeviate(self.rng)
+                use_index = int(real_galaxy_catalogs[0].nobjects * uniform_deviate())
+            else:
+                raise AttributeError('No method specified for selecting a galaxy!')
+            if logger:
+                logger.debug('ChromaticRealGalaxy %d: Start ChromaticRealGalaxy constructor.',
+                             use_index)
+            self.index = use_index
+
+            # Read in the galaxy, PSF images; for now, rely on pyfits to make I/O errors.
+            imgs = [rgc.getGalImage(use_index) for rgc in real_galaxy_catalogs]
+            if logger:
+                logger.debug('ChromaticRealGalaxy %d: Got gal_image', use_index)
+
+            PSFs = [rgc.getPSF(use_index) for rgc in real_galaxy_catalogs]
+            if logger:
+                logger.debug('ChromaticRealGalaxy %d: Got psf', use_index)
+
+            bands = [rgc.getBandpass() for rgc in real_galaxy_catalogs]
+
+            xis = []
+            for rgc in real_galaxy_catalogs:
+                noise_image, pixel_scale, var = rgc.getNoiseProperties(use_index)
+                # Make sure xi image is odd-sized.
+                if noise_image.array.shape[0] % 2 == 0:
```

Comment:
The False branch here never gets run.  I guess all the noise images in the RGCs have even dimensions?  If so, maybe just put a `# pragma: no branch` here with a comment that all of ours are odd, but we don't require that for user-provided RGCs.